Fix several bugs

fgunbin opened the pull request:
* tramp-hdfs-tests.el (hdfs-test-expand-file-name1): Don't assume root
as default user.
* tramp-hdfs.el (tramp-hdfs-file-name-handler-alist): file-readable-p
enters infinite loop, unhandled-file-name-directory refers to
non-existing function.
(tramp-hdfs-handle-file-readable-p): Remove.
(tramp-hdfs-do-rest-call): Remove "--negotiate" (SPNEGO) and "-u :".
(tramp-hdfs-handle-expand-file-name): Fix invocation of
tramp-make-tramp-file-name.
(tramp-hdfs-create-url): Fix invocation of tramp-file-name-host.
(tramp-hdfs-create-one-line): Replace non-existing tramp-half-a-year.
(tramp-hdfs-get-filename): Remove.
(tramp-hdfs-handle-file-local-copy, tramp-hdfs-list-directory)
(tramp-hdfs-handle-delete-directory, tramp-hdfs-handle-delete-file):
Remove its usages.

Commit:
Fix several bugs

* tramp-hdfs-tests.el (hdfs-test-expand-file-name1): Don't assume root
as default user.
* tramp-hdfs.el (tramp-hdfs-file-name-handler-alist): file-readable-p
enters infinite loop, unhandled-file-name-directory refers to
non-existing function.
(tramp-hdfs-handle-file-readable-p): Remove.
(tramp-hdfs-do-rest-call): Remove "--negotiate" (SPNEGO) and "-u :".
(tramp-hdfs-handle-expand-file-name): Fix invocation of
tramp-make-tramp-file-name.
(tramp-hdfs-create-url): Fix invocation of tramp-file-name-host.
(tramp-hdfs-create-one-line): Replace non-existing tramp-half-a-year.
(tramp-hdfs-get-filename): Remove.
(tramp-hdfs-handle-file-local-copy, tramp-hdfs-list-directory)
(tramp-hdfs-handle-delete-directory, tramp-hdfs-handle-delete-file):
Remove its usages.

fgunbin commented:
This PR makes tramp-hdfs operable on current Emacs master.  Tests pass (except for checks requiring specific files to be present on hdfs).
Comments are welcome.

raghavgautam commented on tramp-hdfs.el:
```diff
@@ -237,11 +228,11 @@ Optional argument ARGS is a list of arguments to pass to the OPERATION."
 			      (with-current-buffer
 				  standard-output
-				(call-process tramp-hdfs-curl-path nil t nil "-k" "-sSL" "--negotiate" "-u" ":"  "-X" method url)))))
+				(call-process tramp-hdfs-curl-path nil t nil "-k" "-sSL" "-X" method url)))))
```
How will this work with secure hdfs ?

fgunbin replied:
I think the proper thing to do here is to extract options to a separate parameter - will do it if you don't mind.
I didn't test with secure hdfs, but had some problem with simple local setup - first the macports' curl is not built with SPNEGO (and doesn't have --negotiate), second, I couldn't connect to non-secure hdfs even using curl with spnego.
So it seems logical to allow a user to customize curl invocation.

raghavgautam commented on tramp-hdfs-tests.el:
```diff
@@ -14,13 +14,15 @@
 (ert-deftest hdfs-test-expand-file-name1 ()
   "Tests the expand-file-name for hdfs."
-  (should (equal (expand-file-name "/hdfs:node-1:"             "/tmp") "/hdfs:root@node-1:/"))
+  (should (equal (expand-file-name "/hdfs:node-1:"             "/tmp") (format "/hdfs:%s@node-1:/" (user-login-name))))
```
Can you make this change in other places as well ?

fgunbin replied:
In other places there's explicit user (root), but here it's omitted, so Tramp uses current user

raghavgautam commented:
How did you test this patch ?

raghavgautam commented on tramp-hdfs-tests.el:
```diff
@@ -14,13 +14,15 @@
 
 ;;; Commentary:
 ;; These are some tests for tramp-hdfs.
+;; "node-1" should be mapped to running HDFS instance via /etc/hosts.
```
Can you also put this in running tests section of README ?

fgunbin replied:
sure

fgunbin commented:
> How did you test this patch ?

By running tests and using the package (not very intensively at the moment, I should admit).

fgunbin commented:
I'll get back to this the next week, if that's ok.

Commit:
More fixes

* README.md: Add instructions for running tests.
* tramp-hdfs.el (tramp-hdfs-curl-program, tramp-hdfs-curl-options):
New defcustoms.
(tramp-hdfs-do-rest-call): Use them.

fgunbin commented:
@raghavgautam I've now pushed the changes here.

raghavgautam commented:
@fgunbin Thanks for making the changes. I will review this patch in a few days.

fgunbin commented:
Great, will wait for you feedback.

Commit:
Fix file/directory deletion functions

* tramp-hdfs.el (tramp-hdfs-handle-delete-directory)
(tramp-hdfs-handle-delete-file): Correctly invoke
tramp-flush-file-properties / tramp-flush-directory-properties.

raghavgautam commented:
@fgunbin Thanks for contributing !!

fgunbin commented:
Thank you!
